### PR TITLE
Add partition to the system part_log and merge table.

### PR DESCRIPTION
**Pull request by incfly**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
https://github.com/ClickHouse/ClickHouse/issues/48720. @kgoralski helped with some thought about `system.merges` part. :D  

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
A test needed.

**Comment by alexey-milovidov** [comment]
Now it has a nullptr dereference.

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit a649168ed82d216ff582110d5bfadfb8a8506d6c with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48990/a649168ed82d216ff582110d5bfadfb8a8506d6c/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/MergeTree/MergeList.cpp [hunk 1]
```diff
@@ -40,6 +39,9 @@ MergeListElement::MergeListElement(
     {
         source_data_version = future_part->parts[0]->info.getDataVersion();
         is_mutation = (result_part_info.getDataVersion() != source_data_version);
+
+        WriteBufferFromString out(partition);
+        future_part->getPartition().serializeText(future_part->parts[0]->storage, out, {});
```
Comment (incfly) [inline src/Storages/MergeTree/MergeList.cpp:44]: Looks like this line is causing the exception. https://pastila.nl/?02aa875b/eb263ae20253222b7e4c9127675aa758

Comment (incfly) [reply src/Storages/MergeTree/MergeList.cpp:44]:
```
Cannot parse expression of type Nullable(Int8) here: !Invalid                                                
: While executing ValuesBlockInputFormat
```

And the column right after `partition` is  `{"is_mutation", std::make_shared<DataTypeUInt8>()},` . This makes me suspect somewhere system.merges table schema is not aligned. causing the read results retrying to parse `partition` as `is_mutation`.

Comment (incfly) [reply src/Storages/MergeTree/MergeList.cpp:44]: Ah, we have another place defines the system.merges table. https://github.com/ClickHouse/ClickHouse/blob/55d228e78e977cf9c03d0a5dd922ade2f838b04a/tests/queries/0_stateless/02117_show_create_table_system.reference#L362

Don't understand what's it for, but I figured I should add partition column as well..

File: src/Storages/MergeTree/MergeTreePartition.cpp [hunk 2]
```diff
@@ -371,7 +371,12 @@ void MergeTreePartition::serializeText(const MergeTreeData & storage, WriteBuffe
     const auto & partition_key_sample = metadata_snapshot->getPartitionKey().sample_block;
     size_t key_size = partition_key_sample.columns();
 
-    if (key_size == 0)
+    // Sometimes we can merge empty parts. In that case, the value is empty.
```
Comment (incfly) [inline src/Storages/MergeTree/MergeTreePartition.cpp:374]: Found such log (with debugging instrumentation) when running clickhouse test locally, https://pastila.nl/?0044c8e7/875fee24459b31fc63a2c462f130f45c

Comment (incfly) [inline src/Storages/MergeTree/MergeTreePartition.cpp:374]: I add some instrumentation in the `serializeText` (this function), and then run the fast test via clickhouse-test. In the end, it turns out sometimes ClickHouse is merging two empty parts. https://pastila.nl/?0042a998/918ca76ad1a0ea1bf6e30d9db583b5d3

**Comment by alexey-milovidov** [comment]
Please continue.

File: src/Storages/MergeTree/MergeTreePartition.cpp [hunk 3]
```diff
@@ -379,6 +385,7 @@ void MergeTreePartition::serializeText(const MergeTreeData & storage, WriteBuffe
     {
         const DataTypePtr & type = partition_key_sample.getByPosition(0).type;
         auto column = type->createColumn();
+        const auto & all_column_names = metadata_snapshot->getColumns().getAll();
```
Comment (incfly) [inline src/Storages/MergeTree/MergeTreePartition.cpp:388]: this line is not needed. will remove.

**Comment by incfly** [comment]
```
sed: -e expression #1, char 0: no previous regular expression
Traceback (most recent call last):
  File "/home/ubuntu/actions-runner/_work/_temp/upgrade_memory/ClickHouse/tests/ci/upgrade_check.py", line 4, in <module>
    stress_check.run_stress_test("clickhouse/upgrade-check")
  File "/home/ubuntu/actions-runner/_work/_temp/upgrade_memory/ClickHouse/tests/ci/stress_check.py", line 213, in run_stress_test
    subprocess.check_call(
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'sed -i -r -e 's!!CLICKHOUSE_CI_LOGS_HOST!g; s!!CLICKHOUSE_CI_LOGS_PASSWORD!g;' '/home/ubuntu/actions-runner/_work/_temp/upgrade_memory/run.log'' returned non-zero exit status 1.
```
https://github.com/ClickHouse/ClickHouse/blob/a92fe25ff9968a2edd51f918802c4485957f989a/tests/ci/ast_fuzzer_check.py#L149

The test all passed. https://github.com/ClickHouse/ClickHouse/actions/runs/5872784778/job/15925776487?pr=48990 
I think probably `CLICKHOUSE_CI_LOGS_HOST` is not set, empty?

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**